### PR TITLE
Add type stubs for hvac

### PR DESCRIPTION
**Pull request by kasium**

### CODE REVIEW

**Comment by kasium** [comment]
Not sure about this error

```
error: hvac.v1.Client.__init__ is inconsistent, runtime argument "adapter" has a default value of type abc.ABCMeta, which is incompatible with stub argument type Type[hvac.adapters.Adapter]
Stub: in file /home/runner/work/typeshed/typeshed/stubs/hvac/hvac/v1/__init__.pyi:9
def (self: hvac.v1.Client, url: Union[builtins.str, None] =, token: Union[builtins.str, None] =, cert: Union[tuple[builtins.str, builtins.str], None] =, verify: Union[builtins.bool, builtins.str, None] =, timeout: builtins.int =, proxies: Union[builtins.dict[builtins.str, builtins.str], None] =, allow_redirects: builtins.bool =, session: Union[Any, None] =, adapter: Type[hvac.adapters.Adapter] =, namespace: Union[Any, None] =, **kwargs: Any)
Runtime: in file /tmp/tmp4a7qomfo/lib/python3.11/site-packages/hvac/v1/__init__.py:[69](https://github.com/python/typeshed/actions/runs/8266291204/job/22613913826?pr=11591#step:5:70)
def (self, url=None, token=None, cert=None, verify=None, timeout=30, proxies=None, allow_redirects=True, session=None, adapter=<class 'hvac.adapters.JSONAdapter'>, namespace=None, **kwargs)
```

**Comment by srittau** [comment]
> Not sure about this error

Looks like a weird stubtest hiccup. Add `hvac.v1.Client.__init__` to `stubs/hvac/@tests/stubtest_allowlist.txt` (with an appropriate comment) should work around this.

**Comment by AlexWaygood** [comment]
@kasium, could you run [stubdefaulter](https://github.com/JelleZijlstra/stubdefaulter) on this PR?

**Comment by kasium** [comment]
> @kasium, could you run [stubdefaulter](https://github.com/JelleZijlstra/stubdefaulter) on this PR? We prefer to include simple defaults in stubs nowadays, as they're useful for IDEs

I was not aware of this change in typeshed. Thanks for pointing this out. I added the default values with the help of the tool

**Comment by AlexWaygood** [comment]
> > @kasium, could you run [stubdefaulter](https://github.com/JelleZijlstra/stubdefaulter) on this PR? We prefer to include simple defaults in stubs nowadays, as they're useful for IDEs
> 
> I was not aware of this change in typeshed. Thanks for pointing this out. I added the default values with the help of the tool

Thanks very much!

**Comment by JelleZijlstra** [comment]
Somebody needs to review it. The size of the PR might make some of us shy away from that task.

Also, could you comment on the new error that mypy-primer found? Is it expected or does it indicate a bug in the PR?

File: stubs/hvac/@tests/stubtest_allowlist.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# stubtest does not understand that the adapter argument type
```
Comment (JelleZijlstra) [inline stubs/hvac/@tests/stubtest_allowlist.txt:1]: Incomplete sentence

File: stubs/hvac/hvac/constants/approle.pyi [hunk 2]
```diff
@@ -0,0 +1,4 @@
+from collections.abc import Iterable
+
+DEFAULT_MOUNT_POINT: str
+ALLOWED_TOKEN_TYPES: Iterable[str]
```
Comment (JelleZijlstra) [inline stubs/hvac/hvac/constants/approle.pyi:4]: Iterable is a strange type for a constant. What is it at runtime?

**Review by JelleZijlstra — COMMENTED** [review]
I skimmed through the entire diff and mostly it looks good; many of the files are untyped and stubtest should catch any issues. The last few files have more annotations and could be reviewed in more detail.

File: stubs/hvac/hvac/constants/approle.pyi (hunk 2 above)
Comment (kasium) [reply stubs/hvac/hvac/constants/approle.pyi:4]: Just a list of strings

**Comment by kasium** [comment]
> Somebody needs to review it. The size of the PR might make some of us shy away from that task.
Understood. It's just a big library and I basically ran stubgen 😅

> Also, could you comment on the new error that mypy-primer found? Is it expected or does it indicate a bug in the PR?
Not sure about it. It's weird that mypy complains about this since the module exists. Do you have idea?

**Comment by kasium** [comment]
The mypy_primer error can be reproduced with
```py
import hvac
hvac.exceptions.VaultError
```

**Comment by srittau** [comment]
> The mypy_primer error can be reproduced with
> 
> ```python
> import hvac
> hvac.exceptions.VaultError
> ```
> 
> I guess it's because of the `__all__` in `stubs/hvac/hvac/__init__.pyi`. Any opinion here?

Technically, paasta is incorrect here. `hvac.exceptions` needs to be imported explicitly. But due to Python's side-effect-riddled import mechanism, it's probably somehow imported implicitly. I think keeping the stubs as is is fine.

**Review by srittau — CHANGES_REQUESTED** [review]
Skimming through it, this mostly looks good. Any kinks can be sorted out later. Just a request to remove redundant "Incomplete" annotations: Could you do the following replacements on all code files:

* ` -> Incomplete:` -> `:`
* `: Incomplete)` -> `)`
* `: Incomplete,` -> `,`

**Comment by kasium** [comment]
> Skimming through it, this mostly looks good. Any kinks can be sorted out later. Just a request to remove redundant "Incomplete" annotations: Could you do the following replacements on all code files:
> 
> * ` -> Incomplete:` -> `:`
> * `: Incomplete)` -> `)`
> * `: Incomplete,` -> `,`
> 

Can you please give me some bachground when Incompelte should be used and when not?

**Comment by kasium** [comment]
@srittau done

**Review by srittau — APPROVED** [review]

**Pull request merged**